crates/bevy_ecs/src/schedule/stepping.rs

```diff
@@ -0,0 +1,1305 @@
+        // If it's not present, it will be inserted in place after the index
+        // of the previous schedule that called this function.
+        let index = self.schedule_order.iter().position(|l| l == label);
+        let index = match (index, self.previous_schedule) {
+            (Some(index), _) => index,
+            (None, None) => {
+                self.schedule_order.insert(0, label.clone());
+                0
+            }
+            (None, Some(last)) => {
+                self.schedule_order.insert(last + 1, label.clone());
+                last + 1
+            }
+        };
+        self.previous_schedule = Some(index);
+        println!(
+            "Stepping::skipped_systems(): cursor {:?}, label {:?}, index {}",
+            self.cursor, label, index
+        );
+
+        // if the stepping frame cursor is pointing at this schedule, we'll run
+        // the schedule with the current stepping action.  If this is not the
+        // cursor schedule, we'll run the schedule with the waiting action.
+        let Cursor {
+            schedule: schedule_index,
+            system: start,
+        } = self.cursor;
+        let (list, next_system) = if index == schedule_index {
+            let o = state.skipped_systems(schedule, start, self.action);
+
+            // if we just stepped this schedule, then we'll switch the action
+            // to be waiting
+            if self.action == Action::Step {
+                self.action = Action::Waiting;
+            }
+            o
+        } else {
+            // we're not supposed to run any systems in this schedule, so pull
+            // the skip list, but ignore any changes it makes to the cursor.
+            let (skip, _) = state.skipped_systems(schedule, 0, Action::Waiting);
+            (skip, Some(start))
+        };
+
+        // update the stepping frame cursor based on if there are any systems
+        // remaining to be run in the schedule
+        // Note: Don't try to detect the end of the render frame here using the
+        // schedule index.  We don't know all schedules have been added to the
+        // schedule_order, so only next_frame() knows its safe to reset the
+        // cursor.
+        match next_system {
+            Some(i) => self.cursor.system = i,
+            None => {
+                let index = schedule_index + 1;
+                println!(
+                    "Stepping::skipped_systems(): next schedule {} -> {}",
+                    schedule_index, index,
+                );
+                self.cursor = Cursor {
+                    schedule: index,
+                    system: self.first_system_index_for_schedule(index),
+                }
+            }
+        }
+
+        Some(list)
+    }
+}
+
+#[derive(Default)]
+struct ScheduleState {
+    /// per-system [`SystemBehavior`]
+    behaviors: HashMap<NodeId, SystemBehavior>,
+
+    /// order of NodeIds in the schedule
+    node_ids: Vec<NodeId>,
```

Comment:
Expand comment:
This is a cached copy of `SystemExecutable.system_ids`.  We need it available here to be accessed by `Stepping::cursor()` so we can return `NodeId`s to the caller.